File: java/lowestCommonAncestorOfABinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        //keep track of whether we've found the first target
        boolean foundFirst = false;
        //keep a hash set of the nodes we've found so far
        Set<Integer> foundNodes = new HashSet<>();
        //once we find the first target, keep a copy of the stack at the time, which will be the path from 
        //the root to it
        int[] firstPath = new int[0];
        //the stack will traverse the tree, maintaining the path from the root to the current node
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (stack.size() > 0) {
            TreeNode cur = stack.peek();
            //check if the top of the stack hasn't been found and is one of the targets
            if (!foundNodes.contains(cur.val) && (cur == p || cur == q)) {
                if (!foundFirst) {
                    //if it's the first target, record the path from the root to it
                    foundFirst = true;
                    firstPath = new int[stack.size()];
                    for (int i = 0; i < stack.size(); i++) {
                        firstPath[i] = stack.get(i).val;
                    }
                } else {
                    //if it's the second target, compare the paths to the two targets to find the lowest common ancestor
                    for (int i = 0; i < firstPath.length; i++) {
                        if (i == firstPath.length - 1 || i == stack.size() - 1 || firstPath[i + 1] != stack.get(i + 1).val) {
                            return stack.get(i);
                        }
                    }
                }
            }
            //visit other nodes pre-order traversal
            foundNodes.add(cur.val);
            if (cur.left != null && !foundNodes.contains(cur.left.val)) {
                stack.push(cur.left);
            } else if (cur.right != null && !foundNodes.contains(cur.right.val)) {
                stack.push(cur.right);
            } else {
                stack.pop();
            }
        }
        return root;
    }
}

File: python/lowest_common_ancestor_of_a_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        #keep track of whether we've found the first target
        found_first = False
        #keep a hash set of the nodes we've found so far
        found_nodes = set()
        #once we find the first target, keep a copy of the stack at the time, which will be the path from 
        #the root to it
        first_stack_copy = []
        #the stack will traverse the tree, maintaining the path from the root to the current node
        stack = [root]
        while stack:
            cur = stack[-1]
            #check if the top of the stack hasn't been found and is one of the targets
            if cur not in found_nodes and cur in { p, q }:
                if not found_first:
                    #if it's the first target, record the path from the root to it
                    found_first = True
                    first_stack_copy = stack[:]
                else:
                    #if it's the second target, compare the paths to the two targets to find the lowest common ancestor
                    for i, node in enumerate(first_stack_copy):
                        if i == len(first_stack_copy) - 1 or i == len(stack) - 1 or first_stack_copy[i + 1] != stack[i + 1]:
                            return node
            found_nodes.add(cur)
            #visit other nodes pre-order traversal
            if cur.left is not None and cur.left not in found_nodes:
                stack.append(cur.left)
            elif cur.right is not None and cur.right not in found_nodes:
                stack.append(cur.right)
            else:
                stack.pop()